### PR TITLE
feat: multi-products engine queries

### PR DESCRIPTION
**Pull request by yruej301**
depends on https://github.com/vertex-protocol/vertex-core/pull/864

### CODE REVIEW

**Comment by yruej301** [comment]
@frankfka confirmed all sanity checks are ✅ ✅

File: apps/e2e/src/client/fullSanity.ts [hunk 1]
```diff
@@ -109,6 +109,8 @@ async function fullSanity(context: RunContext) {
 
   // Fetches state from offchain sequencer
   await vertexClient.market.getAllEngineMarkets();
+  await vertexClient.market.getLatestMarketPrices({ productIds: [1, 2, 3] });
+
   // Fetches state from Arbitrum
```
[the remark this answers is not in the document]
Comment (yruej301) [reply apps/e2e/src/client/fullSanity.ts:114]: nope, i will update!

File: packages/engine-client/src/EngineBaseClient.ts [hunk 2]
```diff
@@ -91,20 +91,22 @@ export class EngineBaseClient {
...
     const request = this.getQueryRequest(requestType, params);
-    const queryParams: Record<string, string | number> = {};
+    const requestBody: Record<string, any> = {};
     Object.keys(request).forEach((key) => {
       const value = request[key as keyof typeof request];
       // Remove null values and stringify
       if (value != null) {
-        queryParams[key] = String(value);
+        if (typeof value == 'boolean') {
+          requestBody[key] = String(value);
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineBaseClient.ts:100]: I think backend currently expects strings for booleans -- would require some changes in some of the endpoints, think might be a bit out of scope, but lmk if you feel strongly and i can see what we can do in this pass. fwiw, we have plans for deprecating support for GET on queries at which point we can make this schema changes.

File: apps/e2e/src/engine-client/fullSanity.ts [hunk 3]
```diff
@@ -153,6 +157,13 @@ async function fullSanity(context: RunContext) {
     productId,
   });
   prettyPrint('Queried order', queriedOrder);
+  const queriedOrders = await client.getOrders({
+    subaccountOwner: signer.address,
+    subaccountName: 'default',
+    productIds: [productId, 2, 3],
```
[the remark this answers is not in the document]
Comment (yruej301) [reply apps/e2e/src/engine-client/fullSanity.ts:163]: ye, will fix! just wanted to see the response from multi-product orders query

File: packages/engine-client/src/types/serverQueryTypes.ts [hunk 4]
```diff
@@ -201,6 +217,10 @@ export interface EngineServerMarketPriceResponse {
   ask_x18: string;
 }
 
+export interface EngineServerMarketPricesResponse {
+  market_prices: EngineServerMarketPriceResponse[];
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/types/serverQueryTypes.ts:221]: agree, will fix!

File: packages/engine-client/src/EngineQueryClient.ts [hunk 5]
```diff
@@ -250,6 +254,39 @@ export class EngineQueryClient extends EngineBaseClient {
...
+        subaccountOwner: params.subaccountOwner,
+        subaccountName: params.subaccountName,
+      }),
+      product_ids: params.productIds,
+    });
+
+    const subaccount = subaccountFromHex(baseResponse.sender);
+
+    return {
+      subaccountName: subaccount.subaccountName,
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts:275]: ah, think just kept it for consistency with the existing subaccount orders endpoint -- we can totally remove, will fix!

File: packages/engine-client/src/EngineQueryClient.ts [hunk 6]
```diff
@@ -250,6 +254,39 @@ export class EngineQueryClient extends EngineBaseClient {
     };
   }
 
+  /**
+   * Get all subaccount orders from the engine, for multiple products
+   * @param params
+   */
+  async getOrders(
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts:261]: I think this idea of this endpoint is to eventually become a generic get open orders (similar to get historical orders on the indexer), and in that world, the current name makes sense. We can deff make the naming changes on the SDK.

re: removing `getSubaccountOrders`, think eventually we can -- but believe for now, might be a breaking change for whoever is using it no? (also tbh, i expect the new endpoint to be mostly used by the FE)

tldr: will make the naming changes on the SDK for now and try to minimize breaking changes i.e: think we should keep getSubaccountOrders

File: packages/engine-client/src/EngineBaseClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineBaseClient.ts:100]: actually -- the actual issue here is that the following queries expect strings for booleans (not that the backend can't handle json boolean):

```
MaxOrderSize.spot_leverage

MaxWithdrawable.spot_leverage

MaxLpMintable.spot_leverage
```
think the reason is so that it works using GET via query params. for now, we can just update the sdk to serialize those correctly and we wouldn't need this check anymore. will fix.

File: apps/e2e/src/client/fullSanity.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply apps/e2e/src/client/fullSanity.ts:113]: fixed!

File: packages/engine-client/src/EngineBaseClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineBaseClient.ts]: we can remove it here (see last commit); we still need it in backend to support `GET`

File: apps/e2e/src/engine-client/fullSanity.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply apps/e2e/src/engine-client/fullSanity.ts]: fixed!

File: packages/engine-client/src/types/serverQueryTypes.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/types/serverQueryTypes.ts]: fixed!

File: packages/engine-client/src/EngineQueryClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts]: fixed!

File: packages/engine-client/src/EngineQueryClient.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts]: updated!

File: packages/client/src/apis/market/MarketQueryAPI.ts [hunk 7]
```diff
@@ -50,6 +52,16 @@ export class MarketQueryAPI extends BaseVertexAPI {
     return this.context.engineClient.getSubaccountOrders(params);
   }
 
+  /**
+   * @description Queries the offchain engine to retrieve status of any open orders for the given subaccount for multiple products
+   * @param params
+   */
+  async getOpenSubaccountProductOrders(
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/client/src/apis/market/MarketQueryAPI.ts]: addressed!

File: packages/engine-client/src/types/clientQueryTypes.ts [hunk 8]
```diff
@@ -153,12 +163,22 @@ export interface GetEngineMarketPriceParams {
   productId: number;
 }
 
-export interface GetEngineMarketPriceResponse {
+export interface GetEngineMarketPrice {
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/types/clientQueryTypes.ts]: addressed!

File: packages/engine-client/src/types/serverQueryTypes.ts [hunk 9]
```diff
@@ -72,7 +81,7 @@ export interface EngineServerMaxWithdrawableQueryParams {
   sender: string;
   product_id: number;
   // If not given, engine defaults to true (leverage/borrow enabled)
-  spot_leverage: boolean | null;
+  spot_leverage: string | null;
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/types/serverQueryTypes.ts:84]: believe we decided to leave as is

File: packages/engine-client/src/types/serverQueryTypes.ts [hunk 10]
```diff
@@ -195,13 +213,19 @@ export interface EngineServerSubaccountFeeRatesResponse {
...
 }
 
-export interface EngineServerGetOrderResponse {
+export type EngineServerMarketPriceResponse = EngineServerMarketPrice;
+
+export interface EngineServerMarketPricesResponse {
+  market_prices: EngineServerMarketPrice[];
+}
+
+export interface EngineServerGetOrder {
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/types/serverQueryTypes.ts]: addressed!

File: packages/engine-client/src/EngineQueryClient.ts [hunk 11]
```diff
@@ -326,6 +362,29 @@ export class EngineQueryClient extends EngineBaseClient {
...
+    params: GetEngineMarketPricesParams,
+  ): Promise<GetEngineMarketPricesResponse> {
+    const baseResponse = await this.query('market_prices', {
+      product_ids: params.productIds,
+    });
+    return {
+      marketPrices: baseResponse.market_prices.map(
+        (marketPrice): GetEngineMarketPrice => {
+          return {
+            ask: fromX18(marketPrice.ask_x18),
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts]: addressed!

File: packages/engine-client/src/EngineQueryClient.ts [hunk 12]
```diff
@@ -250,6 +256,35 @@ export class EngineQueryClient extends EngineBaseClient {
...
+    });
+
+    const subaccount = subaccountFromHex(baseResponse.sender);
+
+    return {
+      productOrders: baseResponse.product_orders.map((orders) => {
+        return {
+          orders: orders.orders.map(mapEngineServerOrder),
+          productId: orders.product_id,
+          subaccountOwner: subaccount.subaccountOwner,
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts:281]: this was already there -- i just left it ¯\_(ツ)_/¯

File: packages/engine-client/src/EngineQueryClient.ts [hunk 13]
```diff
@@ -497,3 +546,13 @@ function mapSubaccountSummary(
     },
   };
 }
+
+function mapEngineMarketPrice(
```
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts:550]: ah, will move!

File: apps/e2e/src/client/fullSanity.ts [hunk 14]
```diff
@@ -118,34 +119,39 @@ async function fullSanity(context: RunContext) {
...
   });
+
   await vertexClient.market.getOpenSubaccountOrders({
     subaccountOwner: await signer.getAddress(),
     subaccountName: 'default',
     productId: 3,
   });
-
+  await vertexClient.market.getOpenSubaccountProductOrders({
+    subaccountOwner: await signer.getAddress(),
```
[the remark this answers is not in the document]
Comment (yruej301) [reply apps/e2e/src/client/fullSanity.ts:139]: sg!

File: packages/engine-client/src/EngineQueryClient.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts]: fixed!

File: packages/engine-client/src/EngineQueryClient.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (yruej301) [reply packages/engine-client/src/EngineQueryClient.ts]: fixed!
